pennylane/templates/state_preparations/qrom_state_prep.py

```diff
@@ -0,0 +1,343 @@
+    def _primitive_bind_call(cls, *args, **kwargs):
+        return cls._primitive.bind(*args, **kwargs)
+
+    def _flatten(self):
+        hyperparameters = (
+            ("wires", self.hyperparameters["input_wires"]),
+            ("precision_wires", self.hyperparameters["precision_wires"]),
+            ("work_wires", self.hyperparameters["work_wires"]),
+        )
+        return (self.state_vector,), hyperparameters
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        hyperparams_dict = dict(metadata)
+        return cls(data[0], **hyperparams_dict)
+
+    def map_wires(self, wire_map):
+        new_wires = Wires(
+            [wire_map.get(wire, wire) for wire in self.hyperparameters["input_wires"]]
+        )
+        new_precision_wires = Wires(
+            [wire_map.get(wire, wire) for wire in self.hyperparameters["precision_wires"]]
+        )
+
+        new_work_wires = Wires(
+            [wire_map.get(wire, wire) for wire in self.hyperparameters["work_wires"]]
+        )
+        return QROMStatePreparation(
+            self.state_vector, new_wires, new_precision_wires, new_work_wires
+        )
+
+    def decomposition(self):  # pylint: disable=arguments-differ
+        filtered_hyperparameters = {
+            key: value for key, value in self.hyperparameters.items() if key != "input_wires"
+        }
+        return self.compute_decomposition(
+            self.parameters[0],
+            wires=self.hyperparameters["input_wires"],
+            **filtered_hyperparameters,
+        )
+
+    @staticmethod
+    def compute_decomposition(
+        state_vector, wires, precision_wires, work_wires
+    ):  # pylint: disable=arguments-differ
+        r"""
+        Computes the decomposition operations for the given state vector.
+
+        Args:
+
+            state_vector (TensorLike): The state vector to prepare.
+            wires (Sequence[int]): The wires on which to prepare the state.
+            precision_wires (Sequence[int]): The wires used for storing the binary representations of the
+                amplitudes and phases.
+            work_wires (Sequence[int], optional):  The wires used as work wires for the QROM operations. Defaults to ``None``.
+
+        Returns:
+            list: List of decomposition operations.
+        """
+
+        probs = qml.math.abs(state_vector) ** 2
+        phases = qml.math.angle(state_vector) % (2 * np.pi)
+
+        decomp_ops = []
+        num_iterations = int(qml.math.log2(len(probs)))
+        for i in range(num_iterations):
+
+            # Calculation of the numerator and denominator of the function f(x) (Eq.5 [arXiv:quant-ph/0208112])
+            prefixes = _get_basis_state_list(n_wires=i)
+            probs_denominator = [_sum_by_prefix(probs, prefix=p) for p in prefixes]
+
+            prefixes_with_zero = _get_basis_state_list(n_wires=i, add_zero=True)
+            probs_numerator = [_sum_by_prefix(probs, prefix=p) for p in prefixes_with_zero]
```

Comment:
Could something like this help here to directly compute the binary-bucket sum result without relying on the binary string prefixes?

```py
num_iterations = 4

vector = np.arange(2 ** 4)
probs = vector.reshape(1, -1)

for itx in range(num_iterations):
    probs1 = probs.sum(axis=1) # this gives probs_denominator
    probs = probs.reshape(int(2**(itx+1)), -1) # this makes prefixes_with_zero + prefixes_with_one
    probs2 = probs.sum(axis=1)[::2] # this gives probs_numerator
    
    print(probs1, probs2)
```